libs/logger/logger.hpp

```diff
@@ -18,35 +22,193 @@ auto operator<<(StreamType &os, const T &object)
+                  boost::optional<LogLevel> log_level,
+                  boost::optional<LogPatterns> patterns);
+
+    /// Get tag.
+    const std::string &getTag() const;
+
+    /// Get config.
+    const LoggerConfig &getConfig() const;
+
+    /// Get child config by tag, if present.
+    boost::optional<std::shared_ptr<LoggerConfigTreeNode>> getChild(
+        const std::string &tag) const;
+
+   private:
+    const std::string tag_;
+    const LoggerConfig config_;
+    std::unordered_map<std::string, std::shared_ptr<LoggerConfigTreeNode>>
+        children_;
+  };
+
+  /// Log levels
+  enum class LogLevel {
+    kTrace,
+    kDebug,
+    kInfo,
+    kWarn,
+    kError,
+    kCritical,
+  };
+
+  /// Logger thread safety
+  enum class LoggerThreadSafety {
+    kSingleThread,
+    kMultiThread,
+  };
+
+  class Logger {
+    public:
+     using Level = LogLevel;
+
+     // --- Constructors and assignment (aka the big 5) ---
+
+     /**
+      * Create a logger corresponding to the root of the given tree config.
+      * @param tree_config - the logger config tree
+      * @param ts - thread safety of the created logger
+      */
+     Logger(std::shared_ptr<const LoggerConfigTreeNode> tree_config,
+            LoggerThreadSafety ts);
+
+     /**
+      * Create a standalone logger without tree config.
+      * @param tag - the tag for logging (aka logger name)
+      * @param config - logger configuration
+      * @param ts - thread safety of the created logger
+      */
+     Logger(std::string tag, LoggerConfig config, LoggerThreadSafety ts);
+
+     Logger(const Logger &other);
+     Logger(Logger &&other);
+     Logger &operator=(const Logger &other);
+     Logger &operator=(Logger &&other);
+
+     virtual ~Logger();
+
+     // --- Logging functions ---
+
+     template <typename... Args>
+     void trace(const std::string &format, const Args &... args) const {
+       log(LogLevel::kTrace, format, args...);
+     }
+
+     template <typename... Args>
+     void debug(const std::string &format, const Args &... args) const {
+       log(LogLevel::kDebug, format, args...);
+     }
+
+     template <typename... Args>
+     void info(const std::string &format, const Args &... args) const {
+       log(LogLevel::kInfo, format, args...);
+     }
+
+     template <typename... Args>
+     void warn(const std::string &format, const Args &... args) const {
+       log(LogLevel::kWarn, format, args...);
+     }
+
+     template <typename... Args>
+     void error(const std::string &format, const Args &... args) const {
+       log(LogLevel::kError, format, args...);
+     }
+
+     template <typename... Args>
+     void critical(const std::string &format, const Args &... args) const {
+       log(LogLevel::kCritical, format, args...);
+     }
+
+     template <typename... Args>
+     void log(Level level,
+              const std::string &format,
+              const Args &... args) const {
+       if (shouldLog(level)) {
```

Comment:
You can get rid of conditional and make logger faster, if you rewrite this with 
```
1& logInternal(...)
```

Reply:
Can be just constexpr and compiler will be able to optimize